src/dispatch/decorators.py

```diff
@@ -19,39 +23,114 @@ def fullname(o):
     return f"{module.__name__}.{o.__qualname__}"
 
 
-def scheduled_project_task(func):
-    """Decorator that sets up a background task function with
-    a database session and exception tracking.
-
-    Each task is executed in a specific project context.
-    """
-
-    @wraps(func)
-    def wrapper(*args, **kwargs):
-        db_session = SessionLocal()
-        metrics_provider.counter("function.call.counter", tags={"function": fullname(func)})
-        start = time.perf_counter()
-
+@contextmanager
+def _session(
+    engine: Engine,
+    is_scoped: bool = False,
+) -> None:
+    """Provide a transactional scope around a series of operations."""
+    session = (
+        sessionmaker(bind=engine)()
+        if not is_scoped
+        else scoped_session(sessionmaker(bind=engine))()
+    )
+    try:
+        yield session
+        session.commit()
+    except Exception:
+        session.rollback()
+        raise Exception from None
+    finally:
+        session.close()
+
+
+def _execute_task_in_project_context(
+    func: Callable,
+    is_scoped: bool = False,
+    *args,
+    **kwargs,
+) -> None:
+    db_session = SessionLocal()
+    metrics_provider.counter("function.call.counter", tags={"function": fullname(func)})
+    start = time.perf_counter()
+
+    def __execute_task_within_session_context(
+        schema_session: _GeneratorContextManager[Session],
+    ) -> None:
+        kwargs["db_session"] = schema_session
+        for project in project_service.get_all(db_session=schema_session):
+            project = schema_session.merge(project)
+            kwargs["project"] = project
+            try:
+                func(*args, **kwargs)
+            except Exception as e:
+                log.exception(e)
+
+    try:
         # iterate for all schema
         for organization in organization_service.get_all(db_session=db_session):
             schema_engine = engine.execution_options(
                 schema_translate_map={None: f"dispatch_organization_{organization.slug}"}
             )
-            schema_session = sessionmaker(bind=schema_engine)()
-            kwargs["db_session"] = schema_session
-            for project in project_service.get_all(db_session=schema_session):
-                kwargs["project"] = project
-                try:
-                    func(*args, **kwargs)
-                except Exception as e:
-                    log.exception(e)
-            schema_session.close()
+            if not is_scoped:
+                with _session(
+                    engine=schema_engine,
+                    is_scoped=False,
+                ) as __session:
+                    __execute_task_within_session_context(__session)
+            else:
+                with _session(
+                    engine=schema_engine,
+                    is_scoped=True,
+                ) as __session:
+                    __execute_task_within_session_context(__session)
+
```

Comment:
The `is_scoped` conditional block here is unnecessary. Pass `is_scoped` directly.